File: jupyter2/ps/typedef/time/ml/time.ml
(** begin my answer *)

(** begin hidden *)
type time = Time of (int * int * float);;

let second_of_time (Time (h, m, s)) =
  let h = float_of_int h in
  let m = float_of_int m in
  3600. *. h +. m *. 60. +. s
;;

let time_diff a b =
  let ta = second_of_time a in
  let tb = second_of_time b in
  ta -. tb
;;

(** end hidden *)
(** end my answer *)

let main () =
  assert (Float.abs (time_diff (Time(12, 0, 0.)) (Time(11, 30, 0.)) -. 1800.) < 1e-6);
  assert (Float.abs (time_diff (Time(23, 59, 59.9)) (Time(0, 0, 0.)) -. 86399.9) < 1e-6);
  Printf.printf "OK\n"
;;

main()

File: jupyter2/ps/typedef/time/go/time.go
package main
import "fmt"

/** begin my answer */

/** begin hidden */
type Time struct {
	hour, minute int
	second float64
}

func secondOfTime (t Time) float64 {
	return float64(t.hour) * 3600 + float64(t.minute) * 60 + t.second
}

func timeDiff (a, b Time) float64 {
	return secondOfTime(a) - secondOfTime(b)
}
/** end hidden */
/** end my answer */
func main() {
	if !((timeDiff(Time{hour: 12, minute: 0, second: 0.0}, Time{hour: 11, minute: 30, second: 0.0}) - 1800.0) < 1e-6) { panic("wrong") }
	if !((timeDiff(Time{hour: 23, minute: 59, second: 59.9}, Time{hour: 0, minute: 0, second: 0.0}) - 86399.9) < 1e-6) { panic("wrong") }
	fmt.Println("OK")
}
